Different use_constexpr definitions in spdlog and fmt causes compile time error in NVCC (and possibly other compilers)

# Description
When calling `logger.trace()` in an .cu file compiled using nvcc, I get the following compile time error:
~~~
spdlog/common.h:351:30: error: call to non-‘constexpr’ function ‘const T* fmt::v9::detail::buffer<T>::data() const [with T = char]’
  351 |     return spdlog::string_view_t{buf.data(), buf.size()};
      |                      ~~~~~~~~^~
~~~

# Cause
The [to_string_view](https://github.com/gabime/spdlog/blob/cedfeeb95f3af11df7d3b1e7e0d3b86b334dc23b/include/spdlog/common.h#L349) utility function is declared as [SPDLOG_CONSTEXPR_FUNC](https://github.com/gabime/spdlog/blob/cedfeeb95f3af11df7d3b1e7e0d3b86b334dc23b/include/spdlog/common.h#L72C19-L72C19). The functions calls [fmt::v9::detail::buffer<T>::data()](https://github.com/gabime/spdlog/blob/cedfeeb95f3af11df7d3b1e7e0d3b86b334dc23b/include/spdlog/fmt/bundled/core.h#L911) which is declared as [FMT_CONSTEXPR](https://github.com/gabime/spdlog/blob/cedfeeb95f3af11df7d3b1e7e0d3b86b334dc23b/include/spdlog/fmt/bundled/core.h#L95). However, FMT_CONSTEXPR has different conditions than SPDLOG_CONSTEXPR_FUNC and is defined to be empty for NVCC, leading to the data() function not declared as constexpr and thus causing the error.

# Fix
Adding the [same condition](https://github.com/gabime/spdlog/blob/cedfeeb95f3af11df7d3b1e7e0d3b86b334dc23b/include/spdlog/fmt/bundled/core.h#L95) to the definition of SPDLOG_CONSTEXPR_FUNC fixes the error. However, the two definitions need to be kept in sync in future changes.

Another potential fix is to declare the to_string_view functions as inline instead of constexpr.

--- a/include/spdlog/common.h
+++ b/include/spdlog/common.h
@@ -70,13 +70,19 @@
     #define SPDLOG_CONSTEXPR constexpr
 #endif
 
-// If building with fmt SPDLOG_CONSTEXPR_FUNC needs to be set the same as FMT_CONSTEXPR
-// to avoid situations where a constexpr function in spdlog could end up calling
-// a non-constexpr function in fmt depending on the compiler
-#ifdef FMT_CONSTEXPR
-    #define SPDLOG_CONSTEXPR_FUNC FMT_CONSTEXPR
-#else
+// If building with std::format, can just use constexpr, otherwise if building with fmt
+// SPDLOG_CONSTEXPR_FUNC needs to be set the same as FMT_CONSTEXPR to avoid situations where
+// a constexpr function in spdlog could end up calling a non-constexpr function in fmt
+// depending on the compiler
+// If fmt determines it can't use constexpr, we should inline the function instead
+#ifdef SPDLOG_USE_STD_FORMAT
     #define SPDLOG_CONSTEXPR_FUNC constexpr
+#else  // Being built with fmt
+    #if FMT_USE_CONSTEXPR
+        #define SPDLOG_CONSTEXPR_FUNC FMT_CONSTEXPR
+    #else
+	#define SPDLOG_CONSTEXPR_FUNC inline
+    #endif
 #endif
 
 #if defined(__GNUC__) || defined(__clang__)

--- a/include/spdlog/common.h
+++ b/include/spdlog/common.h
@@ -70,24 +70,19 @@
     #define SPDLOG_CONSTEXPR constexpr
 #endif
 
-#ifdef __has_feature
-#    define SPDLOG_HAS_FEATURE(x) __has_feature(x)
-#else
-#    define SPDLOG_HAS_FEATURE(x) 0
-#endif
-
-// Check if relaxed C++14 constexpr is supported
-// GCC doesn't allow throw in constexpr until version 6 (bug 67371)
-// This needs to stay synchronized with FMT_CONSTEXPR otherwise we can run into
-// situations where spdlog constexpr functions can try to call non-constexpr
-// fmt functions
-// See https://github.com/gabime/spdlog/issues/2856 where this happens with nvcc
-#if (SPDLOG_HAS_FEATURE(cxx_relaxed_constexpr) || (defined(_MSC_VER) && (_MSC_VER >= 1912)) ||                                                  \
-     (defined(__GNUC__) && __GNUC__ >= 6 && defined(__cplusplus) && __cplusplus >= 201402L)) &&                                            \
-    !defined(__ICL) && !defined(__INTEL_COMPILER) && !defined(__NVCC__)
+// If building with std::format, can just use constexpr, otherwise if building with fmt
+// SPDLOG_CONSTEXPR_FUNC needs to be set the same as FMT_CONSTEXPR to avoid situations where
+// a constexpr function in spdlog could end up calling a non-constexpr function in fmt
+// depending on the compiler
+// If fmt determines it can't use constexpr, we should inline the function instead
+#ifdef SPDLOG_USE_STD_FORMAT
     #define SPDLOG_CONSTEXPR_FUNC constexpr
-#else
-    #define SPDLOG_CONSTEXPR_FUNC inline
+#else  // Being built with fmt
+    #if FMT_USE_CONSTEXPR
+        #define SPDLOG_CONSTEXPR_FUNC FMT_CONSTEXPR
+    #else
+	#define SPDLOG_CONSTEXPR_FUNC inline
+    #endif
 #endif
 
 #if defined(__GNUC__) || defined(__clang__)

--- a/include/spdlog/common.h
+++ b/include/spdlog/common.h
@@ -70,11 +70,19 @@
     #define SPDLOG_CONSTEXPR constexpr
 #endif
 
-#ifndef __has_feature
-#    define __has_feature(x) 0
+#ifdef __has_feature
+#    define SPDLOG_HAS_FEATURE(x) __has_feature(x)
+#else
+#    define SPDLOG_HAS_FEATURE(x) 0
 #endif
 
-#if (__has_feature(cxx_relaxed_constexpr) || (defined(_MSC_VER) && (_MSC_VER >= 1912)) ||                                                  \
+// Check if relaxed C++14 constexpr is supported
+// GCC doesn't allow throw in constexpr until version 6 (bug 67371)
+// This needs to stay synchronized with FMT_CONSTEXPR otherwise we can run into
+// situations where spdlog constexpr functions can try to call non-constexpr
+// fmt functions
+// See https://github.com/gabime/spdlog/issues/2856 where this happens with nvcc
+#if (SPDLOG_HAS_FEATURE(cxx_relaxed_constexpr) || (defined(_MSC_VER) && (_MSC_VER >= 1912)) ||                                                  \
      (defined(__GNUC__) && __GNUC__ >= 6 && defined(__cplusplus) && __cplusplus >= 201402L)) &&                                            \
     !defined(__ICL) && !defined(__INTEL_COMPILER) && !defined(__NVCC__)
     #define SPDLOG_CONSTEXPR_FUNC constexpr

--- a/include/spdlog/common.h
+++ b/include/spdlog/common.h
@@ -70,19 +70,13 @@
     #define SPDLOG_CONSTEXPR constexpr
 #endif
 
-// If building with std::format, can just use constexpr, otherwise if building with fmt
-// SPDLOG_CONSTEXPR_FUNC needs to be set the same as FMT_CONSTEXPR to avoid situations where
-// a constexpr function in spdlog could end up calling a non-constexpr function in fmt
-// depending on the compiler
-// If fmt determines it can't use constexpr, we should inline the function instead
-#ifdef SPDLOG_USE_STD_FORMAT
+// If building with fmt SPDLOG_CONSTEXPR_FUNC needs to be set the same as FMT_CONSTEXPR
+// to avoid situations where a constexpr function in spdlog could end up calling
+// a non-constexpr function in fmt depending on the compiler
+#ifdef FMT_CONSTEXPR
+    #define SPDLOG_CONSTEXPR_FUNC FMT_CONSTEXPR
+#else
     #define SPDLOG_CONSTEXPR_FUNC constexpr
-#else  // Being built with fmt
-    #if FMT_USE_CONSTEXPR
-        #define SPDLOG_CONSTEXPR_FUNC FMT_CONSTEXPR
-    #else
-	#define SPDLOG_CONSTEXPR_FUNC inline
-    #endif
 #endif
 
 #if defined(__GNUC__) || defined(__clang__)

--- a/include/spdlog/common.h
+++ b/include/spdlog/common.h
@@ -65,15 +65,21 @@
 #if defined(_MSC_VER) && (_MSC_VER < 1900)
     #define SPDLOG_NOEXCEPT _NOEXCEPT
     #define SPDLOG_CONSTEXPR
-    #define SPDLOG_CONSTEXPR_FUNC inline
 #else
     #define SPDLOG_NOEXCEPT noexcept
     #define SPDLOG_CONSTEXPR constexpr
-    #if __cplusplus >= 201402L
-        #define SPDLOG_CONSTEXPR_FUNC constexpr
-    #else
-        #define SPDLOG_CONSTEXPR_FUNC inline
-    #endif
+#endif
+
+#ifndef __has_feature
+#    define __has_feature(x) 0
+#endif
+
+#if (__has_feature(cxx_relaxed_constexpr) || (defined(_MSC_VER) && (_MSC_VER >= 1912)) ||                                                  \
+     (defined(__GNUC__) && __GNUC__ >= 6 && defined(__cplusplus) && __cplusplus >= 201402L)) &&                                            \
+    !defined(__ICL) && !defined(__INTEL_COMPILER) && !defined(__NVCC__)
+    #define SPDLOG_CONSTEXPR_FUNC constexpr
+#else
+    #define SPDLOG_CONSTEXPR_FUNC inline
 #endif
 
 #if defined(__GNUC__) || defined(__clang__)

